### PR TITLE
Dynamic api: Sparql select and construct components, Report Generator

### PR DESCRIPTION
**Pull request by litvinovg**
[Related issue](https://github.com/vivo-project/VIVO/issues/3721)

# What does this pull request do?
Report generator component to produce reports in docx and xlsx from information in triple store queried by SPARQL construct and select components.
New SPARQL Construct component
SPARQL Select component refactored and extended with new functionality.
New ontology properties.

# What's new?
Parameter type default value (dynapi:defaultValue)
Internal parameters (not provided by input, initialized at the start of action with default serialized value)
Json (not a container type) parameter type
Language tag filtering support in Construct and Select SPARQL Queries ( dynapi:languageFiltering ) + test
Select query can produce Json with ResultSetFormatter + test
Parsing accept language header in converter + test
Construct query + tests
Report generator ( dynapi:ReportGenerator ) + tests
Ontology properties
dynapi:template - template file parameter
dynapi:dataSource - json data for the report
dynapi:report  - report file parameter
Renamed DynapiJsonObject to JsonContainer
User permissions check:
Can be defined like that 
<action_individual> dynapi:accessWhiteList  <white_list_individual>  .
<white_list_individual>
  a   dynapi:GroupAccessWhitelist, dynapi:AccessWhitelist;
  dynapi:userGroup <https://vivoweb.org/ontology/vitro-dynamic-api/user-group/curator> .
Individual from last line is from [this file](https://github.com/vivo-project/Vitro/pull/327/commits/b6f214ce3d8db5e5bb3c890bdb85cd4d09249d47#diff-855d6fbeaf4c99d28f8b30cb61c1d5722881bfbc07c59798d25c8b6813cf9e92)

# How should this be tested?
Run tests

# Interested parties
@VIVO-project/vivo-committers

### CODE REVIEW

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/conversion/ParameterConverterTest.java [hunk 1]
```diff
@@ -131,7 +146,7 @@ private ParameterType createArrayType(String arrayMethod, String arrayArgs, bool
 		} else {
 			arrayImplType.setDeserializationConfig(arrayConfig);			
 		}
-		arrayImplType.setName("java.util.ArrayList");
+		arrayImplType.setClassName("java.util.ArrayList");
```
Comment (chenejac) [inline api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/conversion/ParameterConverterTest.java:149]:
```suggestion
		arrayImplType.setClassName(ArrayList.class.getCanonicalName());
```

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg I hope you will find useful some of my comments. Great work.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Parameters.java [hunk 2]
```diff
@@ -70,4 +70,8 @@ public void dereference() {
 		params = null;
 	}
 
+	public Parameter getFirst() {
+		return params.entrySet().iterator().next().getValue();
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Parameters.java]: I agree, it looks unsafe to reuse. I removed that.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlQuery.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		ParameterizedSparqlString pss = new ParameterizedSparqlString();
+		pss.setCommandText(queryText);
+		setLiterals(dataStore, pss);
+		setUris(dataStore, pss);
+		final String preparedQueryString = pss.toString();
+		return preparedQueryString;
+	}
+	
+	protected boolean isValid(DataStore dataStore) {
+		boolean valid = isValid();
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlQuery.java]: Fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/BinaryView.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.implementation.ByteArray;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.types.ImplementationType;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.types.ParameterType;
+
+public class BinaryView {
+
+	public static boolean isByteArray(Parameter param) {
+		ParameterType type = param.getType();
+		ImplementationType implType = type.getImplementationType();
+		String className = implType.getClassName().getCanonicalName();
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/BinaryView.java:15]: Only for debugging purposes

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.components;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+/*import java.io.FileOutputStream;
+import java.io.OutputStream;*/
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java:8]: fixed

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+		assertEquals(OperationResult.ok(), generator.run(dataStore));
+
+		assertTrue(dataStore.contains(REPORT));
+		byte[] reportBytes = BinaryView.getByteArray(dataStore, reportParam);
+		assertTrue(reportBytes.length > templateByteArray.length);
+		/*
+		 * File file = new File(RESOURCES_PATH + "result-" + templatePath); try
+		 * (OutputStream os = new FileOutputStream(file)) { os.write(reportBytes); }
+		 */
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java:80]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/StringPlainLiteralParam.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		this.setName(var);
+		try {
+			ParameterType type = new ParameterType();
+			type.setName(TYPE_NAME);
+			ImplementationType implType = new ImplementationType();
+			type.setImplementationType(implType);
+			implType.setSerializationConfig(getSerializationConfig());
+			implType.setDeserializationConfig(getDeserializationConfig());	
+			implType.setClassName(Literal.class.getCanonicalName());
+			this.setType(type);
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/StringPlainLiteralParam.java:29]: When there is no language tag or datatype specified the literal is assumed to have the datatype xsd:string.

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlConstructQueryIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    public void beforeEach() {
+        storeModel = new OntModelImpl(OntModelSpec.OWL_MEM);
+    }
+    
+    @Test
+    public void test() throws ConfigurationBeanLoaderException, IOException, ConversionException {
+        loadOntology(ontModel);
+        loadModel(ontModel, TEST_ACTION);
+        loadModel(storeModel, TEST_STORE);
+       // servletContext = new ServletContextStub();
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlConstructQueryIntegrationTest.java]: no, fixed

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlConstructQueryTest.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+		sparql.setQueryText(QUERY_OBJ_VAR);
+		Parameter modelParam = ParameterUtils.createModelParameter(MODEL);
+		sparql.setQueryModel(modelParam);
+		Data modelData = new Data(modelParam);
+		ParameterUtils.addStatement(model, "test:resource", "test:property", "alice" );
+		TestView.setObject(modelData, model);
+		dataStore.addData(MODEL, modelData);
+		Parameter strParam = ParameterUtils.createStringLiteralParameter(STR_VAR);
+		sparql.addInputParameter(strParam);
+		//Not enough params
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SparqlConstructQueryTest.java:74]: It explains why sparql.run returns error on the next line

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SolrQueryTest.java [hunk 10]
```diff
@@ -271,7 +266,7 @@ private Parameter createStringParameter(String name) throws Exception {
 		config.setStaticMethod(false);
 		impltype.setDeserializationConfig(config);
 		impltype.setSerializationConfig(config);
-		impltype.setName("java.lang.String");
+		impltype.setClassName("java.lang.String");
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SolrQueryTest.java]: fixed

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+		final Data data = store.getData(REPORT);
+		assertTrue(TestView.getObject(data) != null);
+		final String base64EncodedReport = data.getSerializedValue();
+		assertTrue(!StringUtils.isEmpty(base64EncodedReport));
+		
+		/*
+		 * byte[] reportBytes = Base64.getDecoder().decode(base64EncodedReport); File
+		 * file = new File(RESOURCES_PATH + "integration-test-report" + extension); try
+		 * (OutputStream os = new FileOutputStream(file)) { os.write(reportBytes); }
+		 */
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorIntegrationTest.java]: fixed

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SolrQueryTest.java [hunk 12]
```diff
@@ -271,7 +266,7 @@ private Parameter createStringParameter(String name) throws Exception {
...
 		
 		RDFType rdfType = new RDFType();
 		rdfType.setName("string");
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/SolrQueryTest.java:257]: fixed

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/ValidatorsTest.java [hunk 13]
```diff
@@ -137,11 +137,11 @@ public Data createData(Object input) throws Exception {
     	paramType.setImplementationType(implType);
     	Data data = new Data(param);
     	if (input instanceof String[]) {
-			implType.setName("java.util.ArrayList");
+			implType.setClassName("java.util.ArrayList");
 			TestView.setObject(data, Arrays.asList((String[])input));
     	} else {
     		
-			implType.setName("java.lang.String");
+			implType.setClassName("java.lang.String");
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/ValidatorsTest.java]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+		try {
+			ParameterType type = new ParameterType();
+			type.setName(TYPE_NAME);
+			ImplementationType implType = new ImplementationType();
+			type.setImplementationType(implType);
+			implType.setSerializationConfig(getSerializationConfig());
+			implType.setDeserializationConfig(getDeserializationConfig());	
+			implType.setClassName(Literal.class.getCanonicalName());
+			RDFType rdfType = new RDFType();
+			rdfType.setName("string");
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java [hunk 15]
```diff
@@ -8,7 +8,7 @@
 
 public class RDFType implements Removable {
 	
-	private static final String ANY_URI = "http://www.w3.org/2001/XMLSchema#anyURI";
+	public static final String ANY_URI = "http://www.w3.org/2001/XMLSchema#anyURI";
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java:12]: Maybe. I think we can add them in further PRs

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.components;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/*import java.io.File;
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorIntegrationTest.java]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Action.java [hunk 17]
```diff
@@ -177,4 +184,26 @@ public String getOutputPath() {
...
+		if (user == null) {
+			return false;
+		}
+		if (user.isRootUser()) {
+			return true;
+		}
+		for (AccessWhitelist filter : accessWhitelists) {
+			if (filter.isAuthorized(user)) {
+				return true;
+			}
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Action.java:200]: The idea was to provide not only information if whitelist allowed user to access action, but also presence in what group allowed or not allowed user to access action.
Let me know if you think that it is excessive info and that logging should still be moved from GroupAccessWhitelist to Action.

[the remark this answers is not in the document]
Comment (chenejac) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Action.java:200]: ok, I understand. Lets keep it as it is now.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+		try {
+			ParameterType type = new ParameterType();
+			type.setName(TYPE_NAME);
+			ImplementationType implType = new ImplementationType();
+			type.setImplementationType(implType);
+			implType.setSerializationConfig(getSerializationConfig());
+			implType.setDeserializationConfig(getDeserializationConfig());	
+			implType.setClassName(Literal.class.getCanonicalName());
+			RDFType rdfType = new RDFType();
+			rdfType.setName("langString");
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java:32]:
```suggestion
			rdfType.setName(RDFType.LANG_STRING);
```

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+	}
+	
+	public static String serialize(Literal literal) {
+		String lang = literal.getLanguage();
+		if (StringUtils.isBlank(lang)) {
+			return literal.getLexicalForm();
+		}
+		return "" + literal.getLexicalForm() + "@" + lang;
+		//return "\"" + literal.getLexicalForm() + "\"@" + lang;
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java:78]:
```suggestion

```

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java [hunk 20]
```diff
@@ -2,24 +2,35 @@
...
+import org.apache.jena.datatypes.xsd.impl.RDFLangString;
 
 import edu.cornell.mannlib.vitro.webapp.dynapi.components.Removable;
 import edu.cornell.mannlib.vitro.webapp.utils.configuration.Property;
 
 public class RDFType implements Removable {
 	
-	private static final String ANY_URI = "http://www.w3.org/2001/XMLSchema#anyURI";
+	public static final String ANY_URI = "http://www.w3.org/2001/XMLSchema#anyURI";
+	public static final String LANG_STRING = "langString";
```
Comment (chenejac) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java:13]: why it is not "http://www.w3.org/1999/02/22-rdf-syntax-ns#langString"? ANY_URI is defined as a complete data type url. 
Please, also note that LiteralParamFactory defines a const for http://www.w3.org/1999/02/22-rdf-syntax-ns#langString

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chenejac) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java:8]: I am not sure why, but I can see only changes in ReportGeneratorIntegrationTest.java. Can you please check? Can you perform the same improvement for this file.

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg please check my comments

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java (hunk 18 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java (hunk 19 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/implementation/LangStringLiteralParam.java]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java (hunk 20 above)
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/types/RDFType.java:13]: When we check literal we need full uri, when we create new RDFDataType, XSDDatatype takes only name as an argument. LiteralParamFactory is creating parameters in runtime, so for that it is not important if it is a name with namespace or not, but to define the same types in n3 it seemed to me more convenient to use short names the way they are required for  XSDDatatypes.

File: api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/test/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/ReportGeneratorTest.java]: fixed

**Review by chenejac — APPROVED** [review]

**Comment by chenejac** [comment]
superseded by https://github.com/vivo-project/Vitro/pull/364

**Pull request closed without merging**